designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+silent!) errors.
+
+In order to develop a serializer which respects the conventions for
+Stan data types, one would need to declare conventions for the mapping
+of Rust data types to Stan data
+types. [serde_json](https://github.com/serde-rs/json) would be nice to
+use, but has some incompatibilities (Rust tuple is represented as an
+array, rather than an object).
+
+Moreover, to represent matrices, complex numbers, etc., one would need
+to support types from the crate ecosystem since the standard library
+lacks these -- [nalgebra](https://github.com/dimforge/nalgebra) and
+[num-complex](https://github.com/rust-num/num-complex) are reasonable
+choices, but nonetheless represent decisions to be made!
+
+From a design perspective, this is a great place to defer to the user,
+at least for the moment. A principled approach would involve writing a
+data format for [serde](https://serde.rs/).
+
+### Deserialization
+
+Parsing Stan CSVs to a strongly-typed representation is simple if one
+wishes to simply obtain a matrix of values (or `Vec<Vec<f64>>` if we
+limit ourselves to `std` library types). However, one needs to extract
+the variables from each line, thus, one needs to know the types (and
+their dimensions). A recursive definition of types using `enum`s and
+`struct`s could probably work to represent such a thing in Rust, but
+may not necessarily be particularly ergonomic (i.e. much unavoidable
+boilerplate would be needed to use the resultant type).
+
+Procedural macros, applied to a Stan program stored in a string
+literal in a Rust program, could be used to generate types and a
+parser for Stan CSVs produced said program. However, in order to
+implement such an idea, one would first need to adopt conventions for
+representing Stan data types using Rust data types. This requires
+careful thought and and is something best left for the future, if
+ever.
+
+The current proposal is for `CmdStanOutput` to be capable of returning
+paths to the files and the user parses them however they desire.
+This leaves open the possibility of multiple parsing strategies.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Representing CmdStan arguments/options as a concrete syntax tree is
+potentially brittle. If the CmdStan grammar undergoes radical change,
+this interface will need to change accordingly. However, the CmdStan
+grammar is intended to be quite stable. Moreover, it is not
+necessarily the case that radical changes to the CmdStan grammar could
+be hidden behind something other abstraction.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Other than the direct representation of the CmdStan syntax tree, the
+proposal contains nothing new. Utilizing a concrete representation of
+the syntax tree does have benefits:
+- all outputs handled via single type `CmdStanOutput`
+- elimination of individual structures and methods for each of Stan's
+inference algorithms
```

Comment:
I think eliminating these is probably a good idea, but maybe not one whose time has come. At the moment, the output files of e.g. ADVI should _not_ be treated the same as those produced by sampling, because the first row is the mean of the approximation, not a draw. Therefore, something like picking a random row for initialization or even passing to `stansummary` is subtly different depending on algorithm.

One hopes that cmdstan will eventually have some nicer output conventions, such as putting the mean in its own file, but in the mean time the type safety of preventing you from doing something foolish could be nice.

Reply:
My thought was that with the argument tree cached in the output, special handling could be implemented there (e.g. `match` on the method). A reference to the cached value would be available to the user, thus enabling the same `match`-based approach for their downstream code.

On the other hand, it would be possible to move the common functionality to a trait and have one output type for each method; this would enable generic code for `T: Output`, but would likely require a `Box<dyn Output>`. 

Another enum, mirroring the method variants in the argument tree, would likely be easier to reason about (rather than a boxed trait object). But if you have to `match` on output variant, it's the same in effect to `match` on the input variant.